<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Welcome to Pynome’s documentation! &#8212; Pynome alpha documentation</title>
    
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     'alpha',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body role="document">
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="welcome-to-pynome-s-documentation">
<h1>Welcome to Pynome&#8217;s documentation!<a class="headerlink" href="#welcome-to-pynome-s-documentation" title="Permalink to this headline">¶</a></h1>
<div class="toctree-wrapper compound">
</div>
<span class="target" id="module-pynome.database"></span><div class="section" id="the-genome-database-module">
<h2>The Genome Database Module<a class="headerlink" href="#the-genome-database-module" title="Permalink to this headline">¶</a></h2>
<p>The <strong>Genomedatabase</strong> module consists of two classes:</p>
<dl class="class">
<dt id="pynome.database.GenomeDatabase">
<em class="property">class </em><code class="descclassname">pynome.database.</code><code class="descname">GenomeDatabase</code><span class="sig-paren">(</span><em>download_path</em>, <em>database_path</em><span class="sig-paren">)</span><a class="headerlink" href="#pynome.database.GenomeDatabase" title="Permalink to this definition">¶</a></dt>
<dd><p>Base Genome Database class. Many functions will be overwritten by
the database-specific child classes.
:param path: the sql database path.</p>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last"><strong>This class should not be directly called.</strong>
It must be implemented with a child class to be useful.</p>
</div>
<dl class="method">
<dt id="pynome.database.GenomeDatabase.get_found_genomes">
<code class="descname">get_found_genomes</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pynome.database.GenomeDatabase.get_found_genomes" title="Permalink to this definition">¶</a></dt>
<dd><p>Print all the Genomes in the database to the terminal.</p>
</dd></dl>

<dl class="method">
<dt id="pynome.database.GenomeDatabase.save_genome">
<code class="descname">save_genome</code><span class="sig-paren">(</span><em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#pynome.database.GenomeDatabase.save_genome" title="Permalink to this definition">¶</a></dt>
<dd><p>The internal function that will save a supplied genome to the
sqlite database.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pynome.database.GenomeEntry">
<em class="property">class </em><code class="descclassname">pynome.database.</code><code class="descname">GenomeEntry</code><span class="sig-paren">(</span><em>taxonomic_name</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#pynome.database.GenomeEntry" title="Permalink to this definition">¶</a></dt>
<dd><p>A sqlite handler for the GenomeTable database.</p>
<p>This supposedly will both create the desired sql table, as well as
act as the handler for generating new row entries into said table.
To add a record to the database, an instance of this class must be
initialized with the desired data within. Then that new instance of
the GenomeEntry will be added to the session, and then committed.
Each GenomeEntry has:</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>taxonomic_name</strong> &#8211; The taxonomic name and primary key of
the GenomeEntry.</li>
<li><strong>download_method</strong> &#8211; The Download method. Stored as
<code class="docutils literal"><span class="pre">&lt;method_name&gt;</span> <span class="pre">&lt;database&gt;</span></code></li>
<li><strong>fasta_uri</strong> &#8211; The fa.gz url as a String. Max Chars = 1000.</li>
<li><strong>gff3_uri</strong> &#8211; The gff3.gz uri as a String. Max Chars = 1000.</li>
<li><strong>genome_local_path</strong> &#8211; The local path of this genome as a String.
Max Chars = 1000.</li>
<li><strong>gff3_size</strong> &#8211; The remote size of the gff3.gz file as an Integer.</li>
<li><strong>fasta_size</strong> &#8211; The remote size of the fa.gz file as an Integer.</li>
<li><strong>assembly_name</strong> &#8211; The name of the assembly.</li>
<li><strong>genus</strong> &#8211; The genus of the assembly.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Examples:</th><td class="field-body"></td>
</tr>
</tbody>
</table>
<p>An instance of this class should be created whenever a genome entry
needs to be created or modified.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">newGenome</span> <span class="o">=</span> <span class="n">GenomeEntry</span><span class="p">([</span><span class="n">taxonomic_name</span><span class="p">],</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</pre></div>
</div>
<p>In deployments this will be handled by a wrapper function specific
to the database being examined.</p>
</dd></dl>

</div>
<span class="target" id="module-pynome.ensembl"></span><div class="section" id="ensembl-database">
<h2>Ensembl Database<a class="headerlink" href="#ensembl-database" title="Permalink to this headline">¶</a></h2>
<p>The ensembl database module. A child class of the genome database class,
this module cotains all code directly related to connecting and parsing data
from the ensembl geneome database.</p>
<dl class="class">
<dt id="pynome.ensembl.EnsemblDatabase">
<em class="property">class </em><code class="descclassname">pynome.ensembl.</code><code class="descname">EnsemblDatabase</code><span class="sig-paren">(</span><em>release_version=36</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#pynome.ensembl.EnsemblDatabase" title="Permalink to this definition">¶</a></dt>
<dd><p>The EnsemblDatabase class. This handles finding and downloading
genomes from the ensembl genome database. The database url is:</p>
<blockquote>
<div><code class="docutils literal"><span class="pre">`ftp.ensemblgenomes.org`</span></code></div></blockquote>
<p>It does so by recursively walking the ftp directory. It only collects
those genomes that have a <code class="docutils literal"><span class="pre">*.gff3.gz</span></code> or a <code class="docutils literal"><span class="pre">*.fa.gz</span></code> file.</p>
<dl class="method">
<dt id="pynome.ensembl.EnsemblDatabase.download_genomes">
<code class="descname">download_genomes</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pynome.ensembl.EnsemblDatabase.download_genomes" title="Permalink to this definition">¶</a></dt>
<dd><p>This function takes an list of genome tuples.
The directory structure to fit the files downloaded is as follows:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">Genome</span><span class="o">/</span>
  <span class="p">[</span><span class="n">genus</span><span class="p">]</span><span class="n">_</span><span class="p">[</span><span class="n">species</span><span class="p">]{</span><span class="n">_</span><span class="p">[</span><span class="n">intraspecific_name</span><span class="p">]}</span><span class="o">/</span>
    <span class="p">[</span><span class="n">assembly_name</span><span class="p">]</span><span class="o">/</span>
      <span class="p">[</span><span class="n">genus</span><span class="p">]</span><span class="n">_</span><span class="p">[</span><span class="n">species</span><span class="p">]{</span><span class="n">_</span><span class="p">[</span><span class="n">intraspecific_name</span><span class="p">]}</span><span class="o">-</span><span class="p">[</span><span class="n">assembly_name</span><span class="p">]</span><span class="o">.</span><span class="n">gff3</span>
      <span class="p">[</span><span class="n">genus</span><span class="p">]</span><span class="n">_</span><span class="p">[</span><span class="n">species</span><span class="p">]{</span><span class="n">_</span><span class="p">[</span><span class="n">intraspecific_name</span><span class="p">]}</span><span class="o">-</span><span class="p">[</span><span class="n">assembly_name</span><span class="p">]</span><span class="o">.</span><span class="n">fasta</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pynome.ensembl.EnsemblDatabase.ensembl_line_parser">
<code class="descname">ensembl_line_parser</code><span class="sig-paren">(</span><em>line</em>, <em>top_dir</em><span class="sig-paren">)</span><a class="headerlink" href="#pynome.ensembl.EnsemblDatabase.ensembl_line_parser" title="Permalink to this definition">¶</a></dt>
<dd><p>This function parses one &#8216;line&#8217; at a time retrieved from an
<code class="docutils literal"><span class="pre">ftp.dir()</span></code> command. This line has already been confirmed to
not be a directory.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>top_dir</strong> &#8211; The parent directory.</li>
<li><strong>line</strong> &#8211; An input line, described in detail below.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>An example of one such line:</p>
<blockquote>
<div><code class="docutils literal"><span class="pre">&quot;drwxr-sr-x</span>&#160; <span class="pre">2</span> <span class="pre">ftp</span>&#160;&#160; <span class="pre">ftp</span>&#160;&#160;&#160; <span class="pre">4096</span> <span class="pre">Jan</span> <span class="pre">13</span>&#160; <span class="pre">2015</span> <span class="pre">filename&quot;</span></code></div></blockquote>
<p>The files are consistently named following this pattern:</p>
<blockquote>
<div><code class="docutils literal"><span class="pre">&lt;species&gt;.&lt;assembly&gt;.&lt;_version&gt;.gff3.gz</span></code></div></blockquote>
<p>This line is split by whitespace. For future reference, the indexes
correspond (usually) to:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>+ ``[0]:    the directory information.``
+ ``[1]:    the number of items therein?``
+ ``[2]:    unknown always &#39;ftp&#39;``
+ ``[3]:    unknown always &#39;ftp&#39;``
+ ``[4]:    the filesize in bytes, 4096 is one block``
+ ``[5]:    Month``
+ ``[6]:    Day``
+ ``[7]:    Year``
+ ``[8]:    filename``
</pre></div>
</div>
<p>Either adds a genome, or returns nothing.</p>
</dd></dl>

<dl class="method">
<dt id="pynome.ensembl.EnsemblDatabase.find_genomes">
<code class="descname">find_genomes</code><span class="sig-paren">(</span><em>uri_list</em>, <em>parsing_function=&lt;function EnsemblDatabase.ensembl_line_parser&gt;</em><span class="sig-paren">)</span><a class="headerlink" href="#pynome.ensembl.EnsemblDatabase.find_genomes" title="Permalink to this definition">¶</a></dt>
<dd><p>Private function that handles finding the list of genomes.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>parsing_function</strong> &#8211; This should be a function that reads an
<code class="docutils literal"><span class="pre">ftplib.dir()</span></code> line output. This output should always be a
file, not a directory.</li>
<li><strong>uri_list</strong> &#8211; This should be a list of base URIs to start
the ftp crawler from.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pynome.ensembl.EnsemblDatabase.generate_metadata_uri">
<code class="descname">generate_metadata_uri</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pynome.ensembl.EnsemblDatabase.generate_metadata_uri" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates a URI that will locate the metadata. This URI is of the
form:</p>
<p><code class="docutils literal"><span class="pre">ftp.ensemblgenomes.org/pub/release-36/species_metadata.json</span></code></p>
</dd></dl>

<dl class="method">
<dt id="pynome.ensembl.EnsemblDatabase.generate_uri">
<code class="descname">generate_uri</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pynome.ensembl.EnsemblDatabase.generate_uri" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates the uri strings needed to download the genomes
from the ensembl database.</p>
<p><strong>Returns</strong>: List of Strings of URIs for the ensembl database. eg:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="s1">&#39;pub/fungi/release-36/gff3/&#39;</span><span class="p">,</span>
<span class="s1">&#39;pub/metazoa/release-36/gff3/&#39;</span><span class="p">,</span>
<span class="o">...</span>
</pre></div>
</div>
<p>This is an extremely case-sepcific function.</p>
</dd></dl>

<dl class="attribute">
<dt id="pynome.ensembl.EnsemblDatabase.release_version">
<code class="descname">release_version</code><a class="headerlink" href="#pynome.ensembl.EnsemblDatabase.release_version" title="Permalink to this definition">¶</a></dt>
<dd><p>Release version property. Should be in the form:
<code class="docutils literal"><span class="pre">&quot;release-#&quot;,</span> <span class="pre">&quot;release-36&quot;</span></code></p>
</dd></dl>

</dd></dl>

</div>
<span class="target" id="module-pynome.ftpHelper"></span><div class="section" id="generic-download-helpers">
<h2>Generic Download Helpers<a class="headerlink" href="#generic-download-helpers" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="pynome.ftpHelper.crawl_ftp_dir">
<code class="descclassname">pynome.ftpHelper.</code><code class="descname">crawl_ftp_dir</code><span class="sig-paren">(</span><em>database</em>, <em>top_dir</em>, <em>parsing_function</em><span class="sig-paren">)</span><a class="headerlink" href="#pynome.ftpHelper.crawl_ftp_dir" title="Permalink to this definition">¶</a></dt>
<dd><p>Recursively crawl a target directory. Takes as an input a
target directory and a parsing function. The ftplib.FTP.dir()
function is used to retrieve a directory listing, line by line,
in string format. These are appended to a newly generated list.
Each item in this list is subject to the parsing function.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>ftp_instance</strong> &#8211; An instance of ftplib.FTP()</li>
<li><strong>top_dir</strong> &#8211; The directory from which contents will be retrieved.</li>
<li><strong>parsing_function</strong> &#8211; The function to parse each non-directory
result.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
</div>
<div class="section" id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></li>
<li><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></li>
<li><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></li>
</ul>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="#">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Welcome to Pynome&#8217;s documentation!</a><ul>
<li><a class="reference internal" href="#the-genome-database-module">The Genome Database Module</a></li>
<li><a class="reference internal" href="#ensembl-database">Ensembl Database</a></li>
<li><a class="reference internal" href="#generic-download-helpers">Generic Download Helpers</a></li>
</ul>
</li>
<li><a class="reference internal" href="#indices-and-tables">Indices and tables</a></li>
</ul>
<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="#">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/index.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2017, Tyler Bigg.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.5.6</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.10</a>
      
      |
      <a href="_sources/index.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>